interfaces/builtin/gpio_chardev.go

```diff
@@ -0,0 +1,230 @@
+
+	serviceSuffix := fmt.Sprintf("gpio-chardev-%s", slotName)
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		// snap-gpio-helper export-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStart: fmt.Sprintf("/usr/lib/snapd/snap-gpio-helper export-chardev %q %q %q %q", strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snap-gpio-helper unexport-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStop: fmt.Sprintf("/usr/lib/snapd/snap-gpio-helper unexport-chardev %q %q %q %q", strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snapd.gpio-chardev-setup.target is used for synchronization of
+		// app services and virtual device setup during boot.
+		WantedBy: "snapd.gpio-chardev-setup.target",
+		Before:   "snapd.gpio-chardev-setup.target",
+	}
+	return spec.AddService(serviceSuffix, service)
+}
+
+func (iface *gpioChardevInterface) SystemdConnectedPlug(spec *systemd.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	slotName := slot.Name()
+	slotSnapName := slot.Snap().InstanceName()
+	plugName := slot.Name()
+	plugSnapName := plug.Snap().InstanceName()
+
+	target := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", slotSnapName, slotName)
+	symlink := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", plugSnapName, plugName)
+
+	// Create symlink pointing to exported virtual slot device.
+	serviceSuffix := fmt.Sprintf("gpio-chardev-%s", plugName)
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		ExecStart:       fmt.Sprintf("/bin/sh -c 'mkdir -p %q && ln -s %q %q'", filepath.Dir(symlink), target, symlink),
+		ExecStop:        fmt.Sprintf("/bin/sh -c 'rm -f %q'", symlink),
+		WantedBy:        "snapd.gpio-chardev-setup.target",
+		Before:          "snapd.gpio-chardev-setup.target",
+	}
+	return spec.AddService(serviceSuffix, service)
+}
+
+func (iface *gpioChardevInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	slotSnapName := slot.Snap().InstanceName()
+	plugSnapName := plug.Snap().InstanceName()
+	snippet := "# Allow access to exported gpio chardev lines\n"
+	// Allow access to exported virtual slot device.
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s rwk,\n", slotSnapName, slot.Name())
+	// Allow access to plug-side symlink to exported virtual slot device.
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/ r,\n", plugSnapName)
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/* r,", plugSnapName)
+	spec.AddSnippet(snippet)
+
+	return nil
+}
+
+func (iface *gpioChardevInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Match exported virtual device on special udev tag set by snap-gpio-helper.
+	spec.TagDevice(fmt.Sprintf(`TAG=="snap_%s_interface_gpio_chardev_%s"`, slot.Snap().InstanceName(), slot.Name()))
```

Comment:
Both instance and slot names could use characters which are not allowed in udev tags. And so they need to be changed a little, see udevTag() in interfaces/dev/spec.go. Maybe that helper should be moved to strutil, because you'll also need to access it in the helper.

Reply:
I checked the validation for instance and slot names, and took a closer look at udevTag(). I am slightly confused, what characters that are allowed in instance and slot names but not in udev tags?

Reply:
sorry I forgot that `-` are fine in tags and that we don't need to reverse the tag